oda_api/token.py

```diff
@@ -62,57 +69,125 @@ def decode_oda_token(token: str, secret_key=None, allow_invalid=False) -> dict:
 
     raise RuntimeError()
 
+
 def decode_oauth2_token(token: str):
     # usually comes in cookies['_oauth2_proxy']
     return json.loads(base64.b64decode(token.split(".")[0]+"=").decode())
 
-#TODO: move to dynaconf
-def discover_token(
+
+def get_token_roles(decoded_token):
+    # extract role(s)
+    roles = None
+    if isinstance(decoded_token['roles'], str):
+        roles = decoded_token['roles'].split(',') if 'roles' in decoded_token else []
+        roles[:] = [r.strip() for r in roles]
+    elif isinstance(decoded_token['roles'], list):
+        roles = decoded_token['roles'] if 'roles' in decoded_token else []
+        roles[:] = [r.strip() for r in roles]
+    return roles
+
+
+def rewrite_token(new_token,
+                  token_write_method: TokenLocation = None
+                  ):
+    current_token, discover_method = discover_token_and_method(allow_invalid=True)
+    current_decoded_token = decode_oda_token(current_token, allow_invalid=True)
+    new_decoded_token = decode_oda_token(new_token, allow_invalid=True)
+
+    current_decoded_token_expires_in_s = current_decoded_token['exp'] - time.time()
+    new_decoded_token_expires_in_s = new_decoded_token['exp'] - time.time()
+
+    if new_decoded_token_expires_in_s < current_decoded_token_expires_in_s:
+        logger.info("the new token has will expire before the current one")
+
+    current_decoded_token_roles = get_token_roles(current_decoded_token)
+    new_decoded_token_roles = get_token_roles(new_decoded_token)
+    if set(current_decoded_token_roles) - set(new_decoded_token_roles) != set():
+        logger.info("The new token has a different set of roles than the current one:\n"
+                    f"roles current token: {current_decoded_token_roles}\n"
+                    f"roles new token: {new_decoded_token_roles}")
+
+    if token_write_method is not None:
+        # TODO perhaps use a more appropriate location
+        with open("old-oda-token_" + str(time.time()), 'w') as ft:
+            ft.write(current_token)
+        if discover_method is not None:
+            if discover_method == TokenLocation.ODA_ENV_VAR:
+                environ.pop('ODA_TOKEN', None)
+            elif discover_method == TokenLocation.FILE_CUR_DIR:
+                if os.path.exists(path.join(getcwd(), ".oda-token")):
+                    os.remove(path.join(getcwd(), ".oda-token"))
+            elif discover_method == TokenLocation.FILE_HOME:
+                if os.path.exists(path.join(environ["HOME"], ".oda-token")):
+                    os.remove(path.join(environ["HOME"], ".oda-token"))
+
+        if token_write_method == TokenLocation.ODA_ENV_VAR:
```

Comment:
actually, could be good to have `token_write_methods`, i.e. a tuple. 

To avoid the problem that sometimes the user writes something which will not be used in the discovery, do just that: check that the token is dicovered after it's written.

Reply:
if it's not discovered: revert.

Reply:
> actually, could be good to have `token_write_methods`, i.e. a tuple.
> 
What does it mean?

Reply:
User may want to write to multiple locations at once. I.e. it could be 
`token_write_methods: Union[Tuple[TokenLocation], TokenLocation]`

Reply:
> if it's not discovered: revert.

You mean remove the newly written token? And so keep using the current/old one?

Reply:
By reverting, so keep using the old one, don't we risk to have the newly refreshed token ending up not being used?